torch/hub.py

```diff
@@ -119,29 +121,74 @@ def _read_url(url):
         return r.read().decode(r.headers.get_content_charset('utf-8'))
 
 
-def _validate_not_a_forked_repo(repo_owner, repo_name, branch):
-    # Use urlopen to avoid depending on local git.
-    headers = {'Accept': 'application/vnd.github.v3+json'}
-    token = os.environ.get(ENV_GITHUB_TOKEN)
-    if token is not None:
-        headers['Authorization'] = f'token {token}'
-    for url_prefix in (
-            f'https://api.github.com/repos/{repo_owner}/{repo_name}/branches',
-            f'https://api.github.com/repos/{repo_owner}/{repo_name}/tags'):
-        page = 0
-        while True:
-            page += 1
-            url = f'{url_prefix}?per_page=100&page={page}'
-            response = json.loads(_read_url(Request(url, headers=headers)))
-            # Empty response means no more data to process
-            if not response:
-                break
-            for br in response:
-                if br['name'] == branch or br['commit']['sha'].startswith(branch):
-                    return
+def _validate_branch(repo_owner, repo_name, branch):
+    # Here we try to make sure that the branch isn't a potentially malicious commit.
+    # This is important because in GitHub the download URL
+    # f'https://github.com/{repo_owner}/{repo_name}/archive/{commit_hash}.zip'
+    # may actually exist *even if* commit_hash is from a different repo_owner / fork.
+    # For example this URL exists:
+    # https://github.com/pytorch/vision/archive/8949c7011facf7801fdf077cc3e4ecd8f0940c7e.zip
+    # even though the commit hash doesn't come from pytorch/vision. It actually comes from a fork
+    # https://github.com/NicolasHug/vision/commit/8949c7011facf7801fdf077cc3e4ecd8f0940c7e
+    # So we want avoid downloading code that can come from a potentionally malicious fork.
+
+    allowed_hash_chars = string.hexdigits.lower()
+    if any(c.lower() not in allowed_hash_chars for c in branch):
+        # The above issue only exists for commit hashes. So if we know that ``branch``
+        # isn't a commit hash, we can return early. This should avoid many GitHub API calls
+        return
```

Comment:
Aren't pull request a branch in github parlance? I.e. creating a pull request and passing `pull/12345` as branch name will bypass the check and has a potential of loading random code, isn't it?

Reply:
oh, thanks for looking at this @malfet , I actually wanted to check with you before merging it to make sure it wasn't introducing a breach.

As far as I can tell the culprit here is the download URL `f"https://github.com/{repo_owner}/{repo_name}/archive/{branch}.zip"`

but even if PR are considered to be a branch, the URL wouldn't exist. For example:

```py
torch.hub.load('pytorch/vision:pull/123', 'vgg11', skip_validation=True)
```

would try to download https://github.com/pytorch/vision/archive/pull/123.zip which leads to a 404

Reply:
Thank you for the explanation. Makes sense then.